cmd/goal/clerk.go

```diff
@@ -1010,9 +1010,38 @@ func assembleFile(fname string, printWarnings bool) (program []byte) {
 	return ops.Program
 }
 
-func assembleFileWithMap(fname string, printWarnings bool) ([]byte, logic.SourceMap) {
-	ops := assembleFileImpl(fname, printWarnings)
-	return ops.Program, logic.GetSourceMap([]string{fname}, ops.OffsetToLine)
+func assembleFileWithMap(sourceFile string, outFile string, printWarnings bool) ([]byte, logic.SourceMap, error) {
+	ops := assembleFileImpl(sourceFile, printWarnings)
+	pathToSourceFromSourceMap, err := determinePathToSourceFromSourceMap(sourceFile, outFile)
+	if err != nil {
+		return nil, logic.SourceMap{}, err
+	}
+	return ops.Program, logic.GetSourceMap([]string{pathToSourceFromSourceMap}, ops.OffsetToSource), nil
+}
+
+func determinePathToSourceFromSourceMap(sourceFile string, outFile string) (string, error) {
+	var pathToSourceFromSourceMap string
+	if sourceFile == stdinFileNameValue {
+		pathToSourceFromSourceMap = "stdin"
+	} else {
+		sourceFileAbsolute, err := filepath.Abs(sourceFile)
+		if err != nil {
+			return "", fmt.Errorf("could not determine absolute path to source file '%s': %w", sourceFile, err)
+		}
+		if outFile == stdoutFilenameValue {
+			pathToSourceFromSourceMap = sourceFileAbsolute
+		} else {
+			outFileAbsolute, err := filepath.Abs(outFile)
+			if err != nil {
+				return "", fmt.Errorf("could not determine absolute path to output file '%s': %w", outFile, err)
+			}
+			pathToSourceFromSourceMap, err = filepath.Rel(filepath.Dir(outFileAbsolute), sourceFileAbsolute)
+			if err != nil {
+				return "", fmt.Errorf("could not determine path from source map to source: %w", err)
+			}
+		}
+	}
+	return pathToSourceFromSourceMap, nil
 }
```

Comment:
I would be inclined to perform immediate returns instead of assigning to pathToSourceFromSourceMap, since you never need to actually operate on the result in multiple steps.

But it seems ok either way.

Reply:
Second this, a bit more effort to track where that variable is used (it ultimately isn't outside of returning), but non-blocker.

Reply:
Good point, changed in 43f409068